File: frontend/src/types/task.ts
import type { Project } from './project'
import type { User } from './user'

export type TaskStatus = 'todo' | 'in_progress' | 'review' | 'done'
export type TaskPriority = 'low' | 'medium' | 'high' | 'critical'

export interface Subtask {
  id: number
  task_id: number
  title: string
  done: boolean
  position: number
  created_at: string
  updated_at: string
}

export interface Task {
  id: number
  title: string
  description: string
  status: TaskStatus
  priority: TaskPriority
  project_id: number
  assignee_id: number | null
  due_date: string | null
  created_at: string
  updated_at: string
  project?: Project
  assignee?: User | null
  subtasks?: Subtask[]
  /** From API: caller can manage task fields (owner / manager / admin / staff). */
  caller_can_manage?: boolean
  /** From API: caller may change status (includes executors assigned to the task). */
  caller_can_change_status?: boolean
}

File: backend/internal/models/task.go
package models

import "time"

type TaskStatus string

const (
	StatusTodo       TaskStatus = "todo"
	StatusInProgress TaskStatus = "in_progress"
	StatusReview     TaskStatus = "review"
	StatusDone       TaskStatus = "done"
)

type TaskPriority string

const (
	PriorityLow      TaskPriority = "low"
	PriorityMedium   TaskPriority = "medium"
	PriorityHigh     TaskPriority = "high"
	PriorityCritical TaskPriority = "critical"
)

type Task struct {
	ID          uint         `gorm:"primaryKey" json:"id"`
	Title       string       `gorm:"not null" json:"title"`
	Description string       `json:"description"`
	Status      TaskStatus   `gorm:"default:'todo'" json:"status"`
	Priority    TaskPriority `gorm:"default:'medium'" json:"priority"`
	ProjectID   uint         `gorm:"not null;index" json:"project_id"`
	Project     Project      `gorm:"foreignKey:ProjectID" json:"-"`
	SectionID   *uint        `gorm:"index" json:"section_id"`
	Section     *TaskSection `gorm:"foreignKey:SectionID" json:"section,omitempty"`
	Position    int          `gorm:"not null;default:0;index" json:"position"`
	AssigneeID  *uint        `gorm:"index" json:"assignee_id"`
	Assignee    *User        `gorm:"foreignKey:AssigneeID" json:"assignee,omitempty"`
	DueDate     *time.Time   `json:"due_date"`
	Subtasks    []Subtask    `gorm:"foreignKey:TaskID" json:"subtasks"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`

	// Caller ACL (не хранится в БД; устанавливается в API-ответах для текущего пользователя).
	CallerCanManage       bool `json:"caller_can_manage" gorm:"-"`
	CallerCanChangeStatus bool `json:"caller_can_change_status" gorm:"-"`
}
